File: whisper.cpp-master/examples/whisper.android/app/src/main/java/com/whispercppdemo/intent/IntentClassifier.kt
package com.whispercppdemo.intent

import android.content.Context
import android.util.Log
import org.json.JSONObject
import org.tensorflow.lite.Interpreter
import java.io.FileInputStream
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import java.util.*
import kotlin.math.exp

private const val LOG_TAG = "IntentClassifier"

data class IntentResult(
    val intent: String,
    val confidence: Float,
    val allProbabilities: Map<String, Float>,
    val slots: Map<String, Any> = emptyMap(),
    val slotConfidence: Float = 0f
)

class IntentClassifier(private val context: Context) {
    
    private var intentClassifier: Interpreter? = null
    private var vocabulary: Map<String, Int> = emptyMap()
    private var intentMapping: Map<Int, String> = emptyMap()
    private var specialTokens: Map<String, Int> = emptyMap()
    private var isInitialized = false
    private val slotExtractor = SlotExtractor()
    
    companion object {
        private const val MAX_SEQUENCE_LENGTH = 256  // Updated to match your model
        private const val CLS_TOKEN = "[CLS]"
        private const val SEP_TOKEN = "[SEP]"
        private const val PAD_TOKEN = "[PAD]"
        private const val UNK_TOKEN = "[UNK]"
    }
    
    suspend fun initialize(): Boolean {
        return try {
            Log.d(LOG_TAG, "Initializing Intent Classifier...")
            
            // Load metadata
            loadMetadata()
            
            // Load vocabulary and special tokens
            loadVocabulary()
            loadSpecialTokens()
            
            // Load the complete TFLite model (end-to-end)
            loadIntentClassifier()
            
            isInitialized = true
            Log.d(LOG_TAG, "✅ Intent Classifier initialized successfully")
            true
        } catch (e: Exception) {
            Log.e(LOG_TAG, "❌ Failed to initialize Intent Classifier", e)
            false
        }
    }
    
    private fun loadMetadata() {
        val metadataJson = context.assets.open("model_metadata.json").bufferedReader().use { it.readText() }
        val metadata = JSONObject(metadataJson)
        
        // Load intent mapping
        val intentMappingObj = metadata.getJSONObject("intent_mapping")
        val tempIntentMapping = mutableMapOf<Int, String>()
        
        intentMappingObj.keys().forEach { key ->
            tempIntentMapping[key.toInt()] = intentMappingObj.getString(key)
        }
        
        intentMapping = tempIntentMapping
        Log.d(LOG_TAG, "📋 Loaded ${intentMapping.size} intent mappings")
    }
    
    private fun loadVocabulary() {
        val vocabText = context.assets.open("tokenizer/vocab.txt").bufferedReader().use { it.readText() }
        val tempVocabulary = mutableMapOf<String, Int>()
        
        vocabText.lines().forEachIndexed { index, line ->
            val token = line.trim()
            if (token.isNotEmpty()) {
                tempVocabulary[token] = index
            }
        }
        
        vocabulary = tempVocabulary
        Log.d(LOG_TAG, "📝 Loaded vocabulary with ${vocabulary.size} tokens")
    }
    
    private fun loadSpecialTokens() {
        val tempSpecialTokens = mutableMapOf<String, Int>()
        tempSpecialTokens[CLS_TOKEN] = vocabulary[CLS_TOKEN] ?: 101
        tempSpecialTokens[SEP_TOKEN] = vocabulary[SEP_TOKEN] ?: 102
        tempSpecialTokens[PAD_TOKEN] = vocabulary[PAD_TOKEN] ?: 0
        tempSpecialTokens[UNK_TOKEN] = vocabulary[UNK_TOKEN] ?: 100
        
        specialTokens = tempSpecialTokens
        Log.d(LOG_TAG, "🎯 Loaded special tokens: $specialTokens")
    }
    
    private fun loadIntentClassifier() {
        val model = loadModelFile("intent_classifier.tflite")
        intentClassifier = Interpreter(model)
        
        val inputDetails = intentClassifier!!.inputTensorCount
        val outputDetails = intentClassifier!!.outputTensorCount
        
        Log.d(LOG_TAG, "🎯 Complete intent classifier loaded - Inputs: $inputDetails, Outputs: $outputDetails")
        
        // Log input/output shapes
        for (i in 0 until inputDetails) {
            val shape = intentClassifier!!.getInputTensor(i).shape()
            Log.d(LOG_TAG, "  Input $i shape: ${shape.contentToString()}")
        }
        
        for (i in 0 until outputDetails) {
            val shape = intentClassifier!!.getOutputTensor(i).shape()
            Log.d(LOG_TAG, "  Output $i shape: ${shape.contentToString()}")
        }
    }
    
    private fun loadModelFile(fileName: String): MappedByteBuffer {
        val fileDescriptor = context.assets.openFd(fileName)
        val inputStream = FileInputStream(fileDescriptor.fileDescriptor)
        val fileChannel = inputStream.channel
        val startOffset = fileDescriptor.startOffset
        val declaredLength = fileDescriptor.declaredLength
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength)
    }
    
    fun classifyIntent(text: String): IntentResult? {
        if (!isInitialized) {
            Log.e(LOG_TAG, "❌ Intent classifier not initialized")
            return null
        }

        return try {
            Log.d(LOG_TAG, "🔍 Classifying: '$text'")
            
            // Step 1: Tokenize text using BERT tokenizer
            val tokenization = tokenizeText(text)
            val inputIds = tokenization.first
            val attentionMask = tokenization.second
            
            Log.d(LOG_TAG, "📊 Tokenized - InputIds: ${inputIds.take(10).joinToString()}")
            Log.d(LOG_TAG, "📊 AttentionMask: ${attentionMask.take(10).joinToString()}")
            
            // Step 2: Run the complete end-to-end model
            val logits = runCompleteModel(inputIds, attentionMask)
            
            // Step 3: Apply softmax and get predictions
            val probabilities = softmax(logits)
            val bestIndex = probabilities.indices.maxByOrNull { probabilities[it] } ?: 0
            val bestIntent = intentMapping[bestIndex] ?: "Unknown"
            val confidence = probabilities[bestIndex]
            
            // Create probability map for all intents
            val allProbabilities = mutableMapOf<String, Float>()
            probabilities.forEachIndexed { index, probability ->
                val intent = intentMapping[index] ?: "Unknown_$index"
                allProbabilities[intent] = probability
            }
            
            Log.d(LOG_TAG, "✅ Final result: $bestIntent (confidence: ${"%.3f".format(confidence)})")
            
            // Log top 3 predictions
            val sortedIndices = probabilities.indices.sortedByDescending { probabilities[it] }
            Log.d(LOG_TAG, "🏆 Top 3 predictions:")
            sortedIndices.take(3).forEach { index ->
                val intent = intentMapping[index] ?: "Unknown_$index"
                val prob = probabilities[index]
                Log.d(LOG_TAG, "  ${intent}: ${"%.3f".format(prob)}")
            }
            
            // Step 4: Extract slots for the predicted intent
            val slotResult = slotExtractor.extractSlots(text, bestIntent)
            
            IntentResult(
                intent = bestIntent,
                confidence = confidence,
                allProbabilities = allProbabilities,
                slots = slotResult.slots,
                slotConfidence = slotResult.confidence
            )
        } catch (e: Exception) {
            Log.e(LOG_TAG, "❌ Error classifying intent", e)
            null
        }
    }
    
    private fun tokenizeText(text: String): Pair<IntArray, IntArray> {
        // Preprocess text (lowercase, basic cleaning)
        val processedText = text.lowercase(Locale.getDefault()).trim()
        
        // Basic BERT-like tokenization
        val tokens = mutableListOf<String>()
        tokens.add(CLS_TOKEN)  // Add [CLS] token at the beginning
        
        // Simple word tokenization (you might want to implement subword tokenization)
        val words = processedText.split("\\s+".toRegex())
        for (word in words) {
            if (word.isNotEmpty()) {
                // Simple tokenization - can be improved with subword tokenization
                val cleanWord = word.replace(Regex("[^a-zA-Z0-9]"), "")
                if (cleanWord.isNotEmpty()) {
                    tokens.add(cleanWord)
                }
            }
        }
        
        tokens.add(SEP_TOKEN)  // Add [SEP] token at the end
        
        // Convert tokens to IDs
        val inputIds = IntArray(MAX_SEQUENCE_LENGTH) { specialTokens[PAD_TOKEN] ?: 0 }
        val attentionMask = IntArray(MAX_SEQUENCE_LENGTH) { 0 }
        
        val maxTokens = minOf(tokens.size, MAX_SEQUENCE_LENGTH)
        for (i in 0 until maxTokens) {
            val token = tokens[i]
            inputIds[i] = vocabulary[token] ?: specialTokens[UNK_TOKEN] ?: 100
            attentionMask[i] = 1
        }
        
        Log.d(LOG_TAG, "🔤 Tokenized '${processedText}' -> ${tokens.take(10).joinToString()}")
        Log.d(LOG_TAG, "🔤 Token count: ${tokens.size}, Valid tokens: ${attentionMask.sum()}")
        
        return Pair(inputIds, attentionMask)
    }
    
    private fun runCompleteModel(inputIds: IntArray, attentionMask: IntArray): FloatArray {
        // Prepare inputs for TFLite model
        val inputIdsArray = Array(1) { inputIds }
        val attentionMaskArray = Array(1) { attentionMask }
        
        // Prepare output
        val outputArray = Array(1) { FloatArray(intentMapping.size) }
        
        // Run inference
        val inputs = arrayOf(inputIdsArray, attentionMaskArray)
        val outputs = mapOf(0 to outputArray)
        
        intentClassifier!!.runForMultipleInputsOutputs(inputs, outputs)
        
        val logits = outputArray[0]
        
        Log.d(LOG_TAG, "🎯 Model output logits: ${logits.take(5).map { "%.3f".format(it) }.joinToString()}")
        
        return logits
    }
    
    private fun softmax(logits: FloatArray): FloatArray {
        val maxLogit = logits.maxOrNull() ?: 0f
        val expValues = logits.map { exp(it - maxLogit) }
        val sumExp = expValues.sum()
        return expValues.map { (it / sumExp).toFloat() }.toFloatArray()
    }
    
    fun getIntentList(): List<String> {
        return intentMapping.values.sorted()
    }
    
    fun close() {
        intentClassifier?.close()
        isInitialized = false
        Log.d(LOG_TAG, "🔒 Intent classifier closed")
    }
}

File: examples/whisper.swiftui/whisper.swiftui.demo/Intent/IntentClassifier.swift
import Foundation
import TensorFlowLite
import os.log

// MARK: - HFTokenizer Support

struct TokenizationResult {
    let ids: [Int]
    let attentionMask: [Int]
}

class HFTokenizer {
    private let tokenizerData: Data
    
    init(_ data: Data) {
        self.tokenizerData = data
    }
    
    func tokenize(_ text: String) -> TokenizationResult {
        // This is a simplified implementation
        // In a real implementation, you would use a proper HuggingFace tokenizer library
        // For now, we'll implement basic tokenization
        
        let words = text.lowercased().components(separatedBy: .whitespacesAndNewlines).filter { !$0.isEmpty }
        var ids: [Int] = [101] // [CLS] token
        var attentionMask: [Int] = [1]
        
        for word in words {
            // Simple hash-based token ID generation (replace with proper tokenizer)
            let tokenId = abs(word.hashValue) % 30000 + 1000
            ids.append(tokenId)
            attentionMask.append(1)
        }
        
        ids.append(102) // [SEP] token
        attentionMask.append(1)
        
        return TokenizationResult(ids: ids, attentionMask: attentionMask)
    }
    
    func close() {
        // Cleanup if needed
    }
}

class IntentClassifier: ObservableObject {
    
    // MARK: - Constants
    
    private static let maxSequenceLength = 256  // Updated to match Kotlin version
    private static let logger = Logger(subsystem: "com.whispercpp.demo", category: "IntentClassifier")
    
    // MARK: - Properties
    
    @Published var isInitialized = false
    @Published var errorMessage: String?
    
    private var intentClassifier: Interpreter?
    private var hfTokenizer: HFTokenizer?
    private var intentMapping: [Int: String] = [:]
    private let slotExtractor = SlotExtractor()
    
    // MARK: - Initialization
    
    func initialize() async -> Bool {
        do {
            Self.logger.info("Initializing Intent Classifier...")
            
            // Load label encoder (replaces metadata)
            try await loadLabelEncoder()
            
            // Load HuggingFace tokenizer
            try await loadHFTokenizer()
            
            // Load the complete TFLite model (end-to-end)
            try await loadIntentClassifier()
            
            await MainActor.run {
                self.isInitialized = true
                self.errorMessage = nil
            }
            
            Self.logger.info("Intent Classifier initialized successfully")
            return true
            
        } catch {
            Self.logger.error("Failed to initialize Intent Classifier: \(error.localizedDescription)")
            await MainActor.run {
                self.errorMessage = error.localizedDescription
                self.isInitialized = false
            }
            return false
        }
    }
    
    // MARK: - Model Loading
    
    private func loadLabelEncoder() async throws {
        guard let path = Bundle.main.path(forResource: "label_encoder", ofType: "json"),
              let data = try? Data(contentsOf: URL(fileURLWithPath: path)),
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let labelToIntentDict = json["label_to_intent"] as? [String: String] else {
            throw IntentClassificationError.metadataNotLoaded
        }
        
        var tempIntentMapping: [Int: String] = [:]
        for (key, value) in labelToIntentDict {
            if let intKey = Int(key) {
                tempIntentMapping[intKey] = value
            }
        }
        
        self.intentMapping = tempIntentMapping
        Self.logger.info("Loaded \(intentMapping.count) intent mappings from label_encoder.json")
    }
    
    private func loadHFTokenizer() async throws {
        // Load tokenizer.json file from tokenizer folder
        guard let path = Bundle.main.path(forResource: "tokenizer/tokenizer", ofType: "json"),
              let tokenizerData = try? Data(contentsOf: URL(fileURLWithPath: path)) else {
            throw IntentClassificationError.tokenizerNotLoaded
        }
        
        // Initialize HuggingFace tokenizer
        self.hfTokenizer = HFTokenizer(tokenizerData)
        Self.logger.info("HuggingFace tokenizer loaded from tokenizer/tokenizer.json")
    }
    
    private func loadIntentClassifier() async throws {
        guard let modelPath = Bundle.main.path(forResource: "intent_classifier", ofType: "tflite") else {
            throw IntentClassificationError.modelNotLoaded
        }
        
        var options = Interpreter.Options()
        options.threadCount = 2
        
        self.intentClassifier = try Interpreter(modelPath: modelPath, options: options)
        try self.intentClassifier?.allocateTensors()
        
        let inputDetails = intentClassifier?.inputTensorCount ?? 0
        let outputDetails = intentClassifier?.outputTensorCount ?? 0
        
        Self.logger.info("Complete intent classifier loaded - Inputs: \(inputDetails), Outputs: \(outputDetails)")
        
        // Log input/output shapes
        for i in 0..<inputDetails {
            if let shape = try? intentClassifier?.input(at: i).shape {
                Self.logger.info("  Input \(i) shape: \(shape)")
            }
        }
        
        for i in 0..<outputDetails {
            if let shape = try? intentClassifier?.output(at: i).shape {
                Self.logger.info("  Output \(i) shape: \(shape)")
            }
        }
    }
    
    // MARK: - Intent Classification
    
    func classifyIntent(_ text: String) async -> IntentResult? {
        guard isInitialized else {
            Self.logger.error("Intent classifier not initialized")
            return nil
        }
        
        do {
            Self.logger.info("Classifying: '\(text)'")
            
            // Step 1: Tokenize text using BERT tokenizer
            let tokenization = try await tokenizeText(text)
            let inputIds = tokenization.inputIds
            let attentionMask = tokenization.attentionMask
            
            Self.logger.info("Tokenized - InputIds: \(Array(inputIds.prefix(10)).map { String($0) }.joined(separator: ", "))")
            Self.logger.info("AttentionMask: \(Array(attentionMask.prefix(10)).map { String($0) }.joined(separator: ", "))")
            
            // Step 2: Run the complete end-to-end model
            let logits = try await runCompleteModel(inputIds: inputIds, attentionMask: attentionMask)
            
            // Step 3: Apply softmax and get predictions
            let probabilities = softmax(logits)
            guard let bestIndex = probabilities.enumerated().max(by: { $0.element < $1.element })?.offset else {
                throw IntentClassificationError.predictionFailed
            }
            
            let bestIntent = intentMapping[bestIndex] ?? "Unknown"
            let confidence = probabilities[bestIndex]
            
            // Create probability map for all intents
            var allProbabilities: [String: Float] = [:]
            for (index, probability) in probabilities.enumerated() {
                let intent = intentMapping[index] ?? "Unknown_\(index)"
                allProbabilities[intent] = probability
            }
            
            Self.logger.info("Final result: \(bestIntent) (confidence: \(String(format: "%.3f", confidence)))")
            
            // Log top 3 predictions
            let sortedIndices = probabilities.enumerated().sorted { $0.element > $1.element }.map { $0.offset }
            Self.logger.info("Top 3 predictions:")
            for i in 0..<min(3, sortedIndices.count) {
                let index = sortedIndices[i]
                let intent = intentMapping[index] ?? "Unknown_\(index)"
                let prob = probabilities[index]
                Self.logger.info("  \(intent): \(String(format: "%.3f", prob))")
            }
            
            // Step 4: Extract slots for the predicted intent
            let slotResult = await slotExtractor.extractSlots(text: text, intent: bestIntent)
            
            return IntentResult(
                intent: bestIntent,
                confidence: confidence,
                allProbabilities: allProbabilities,
                slots: slotResult.slots,
                slotConfidence: slotResult.confidence
            )
            
        } catch {
            Self.logger.error("Error classifying intent: \(error.localizedDescription)")
            return nil
        }
    }
    
    // MARK: - Text Processing
    
    private func tokenizeText(_ text: String) async throws -> (inputIds: [Int32], attentionMask: [Int32]) {
        // Use HuggingFace tokenizer for proper BERT WordPiece tokenization
        guard let tokenizer = hfTokenizer else {
            throw IntentClassificationError.tokenizerNotLoaded
        }
        
        let result = tokenizer.tokenize(text)
        
        // Ensure the sequences are exactly maxSequenceLength
        var inputIds = Array(repeating: Int32(0), count: Self.maxSequenceLength)
        var attentionMask = Array(repeating: Int32(0), count: Self.maxSequenceLength)
        
        // Copy tokens up to maxSequenceLength, padding or truncating as needed
        let tokenCount = min(result.ids.count, Self.maxSequenceLength)
        for i in 0..<tokenCount {
            inputIds[i] = Int32(result.ids[i])
            attentionMask[i] = Int32(result.attentionMask[i])
        }
        
        // The rest remain as 0 (padding)
        let textPreview = text.count > 50 ? String(text.prefix(50)) + "..." : text
        Self.logger.info("HF Tokenized '\(textPreview)' -> \(tokenCount) tokens")
        Self.logger.info("First 10 input_ids: \(Array(inputIds.prefix(10)).map { String($0) }.joined(separator: ", "))")
        Self.logger.info("Valid tokens: \(attentionMask.reduce(0, +))")
        
        return (inputIds: inputIds, attentionMask: attentionMask)
    }
    
    private func runCompleteModel(inputIds: [Int32], attentionMask: [Int32]) async throws -> [Float] {
        // Prepare inputs for TFLite model
        let inputIdsData = Data(bytes: inputIds, count: inputIds.count * MemoryLayout<Int32>.size)
        let attentionMaskData = Data(bytes: attentionMask, count: attentionMask.count * MemoryLayout<Int32>.size)
        
        try intentClassifier?.copy(inputIdsData, toInputAt: 0)
        try intentClassifier?.copy(attentionMaskData, toInputAt: 1)
        try intentClassifier?.invoke()
        
        let outputTensor = try intentClassifier?.output(at: 0)
        guard let outputData = outputTensor?.data else {
            throw IntentClassificationError.predictionFailed
        }
        
        let logits = outputData.withUnsafeBytes { bytes in
            Array(bytes.bindMemory(to: Float.self))
        }
        
        Self.logger.info("Model output logits: \(Array(logits.prefix(5)).map { String(format: "%.3f", $0) }.joined(separator: ", "))")
        
        return logits
    }
    
    private func softmax(_ logits: [Float]) -> [Float] {
        let maxLogit = logits.max() ?? 0.0
        let expValues = logits.map { exp($0 - maxLogit) }
        let sumExp = expValues.reduce(0, +)
        return expValues.map { $0 / sumExp }
    }
    
    // MARK: - Utility Methods
    
    func getIntentList() -> [String] {
        return Array(intentMapping.values).sorted()
    }
    
    func cleanup() {
        intentClassifier = nil
        hfTokenizer?.close()
        hfTokenizer = nil
        isInitialized = false
        Self.logger.info("Intent classifier cleaned up")
    }
    
    deinit {
        cleanup()
    }
}